src/nrncvode/sptree.hpp

```diff
@@ -13,123 +13,117 @@
 
 #define STRCMP(a, b) (a - b)
 
-template <typename SPBLK>
-struct SPTREE {
-    SPBLK* root; /* root node */
-
-    int enqcmps; /* compares in spenq */
+template <typename T>
+class SPTree {
+  public:
+    SPTree() = default;
+
+    // Is this SPTree empty.
+    bool empty() const;
+
+    // Return the value of enqcmps.
+    int get_enqcmps() const;
+
+    // Insert item in a tree.
+    //
+    // Put n after all other nodes with the same key; when this is
+    // done, n will be the root of the splay tree, all nodes
+    // with keys less than or equal to that of n will be in the
+    // left subtree, all with greater keys will be in the right subtree;
+    // the tree is split into these subtrees from the top down, with rotations
+    // performed along the way to shorten the left branch of the right subtree
+    // and the right branch of the left subtree
+    void enqueue(T* n);
+
+    // Return and remove first node.
+    //
+    // Remove and return the first node; this deletes
+    // (and returns) the leftmost node, replacing it with its right
+    // subtree (if there is one); on the way to the leftmost node, rotations
+    // are performed to shorten the left branch of the tree.
+    T* dequeue();
+
+    // Return the first element in the tree.
+    //
+    // Returns a reference to the first item;
+    // event, and the old left branch is shortened, as if the tree had
+    // been splayed about the first element; this is done by dequeueing
+    // the head and then making the resulting queue the right son of
+    // the head returned by spdeq.
+    T* first();
```

Comment:
If first the smallest item of the root of the tree? I think previously the name was `head`, which seems clearer. To me `first` sounds a lot like the element with the lowest key (ties broke arbitrarily).

Reply:
neither is very clear. In doubt I would probably stick with `head` just not to change naming and confuse things further. But it is not actually the head of the tree, but rather the element with the lowest key. The Splay tree is here used like a priority queue, we are mostly interested in extracting the min element.
We might want to name this `min` or `top` .

Reply:
This is the item with the lowest key, yes.

Reply:
Then it's a good name, please update the documentation to state as much. Currently, it defines `first` by using "first" a lot :)